EKF/ekf_helper.cpp

```diff
@@ -964,12 +964,109 @@ void Ekf::get_gyro_bias(float bias[3])
 	memcpy(bias, temp, 3 * sizeof(float));
 }
 
+// covariance propagation from quaternions to euler angles using the covariance law
+// source: "Development of a Real-Time Attitude System Using a Quaternion
+// Parameterization and Non-Dedicated GPS Receivers", John B. Schleppe 1996
+const matrix::SquareMatrix<float, 3> Ekf::propagate_covariances_from_quat_to_euler() const
+{
+	// Jacobian matrix (3x4) containing the partial derivatives of the
+	// Euler angle equations with respect to the quaternions
+	matrix::Matrix<float, 3, 4> G;
+
+	float q1 = _output_new.quat_nominal(0);
+	float q2 = _output_new.quat_nominal(1);
+	float q3 = _output_new.quat_nominal(2);
+	float q4 = _output_new.quat_nominal(3);
+
+	G(0,0) = -(q3+q2) / ((q3+q2)*(q3+q2) + (q4+q2)*(q4+q1)) + (q3-q2)/((q3-q2)*(q3-q2) + (q4-q1)*(q4-q1));
+	G(0,1) =  (q4+q1) / ((q3+q2)*(q3+q2) + (q4+q2)*(q4+q1)) - (q4-q1)/((q3-q2)*(q3-q2) + (q4-q1)*(q4-q1));
+	G(0,2) =  (q4+q1) / ((q3+q2)*(q3+q2) + (q4+q2)*(q4+q1)) + (q4-q1)/((q3-q2)*(q3-q2) + (q4-q1)*(q4-q1));
+	G(0,3) = -(q3+q2) / ((q3+q2)*(q3+q2) + (q4+q2)*(q4+q1)) - (q3-q2)/((q3-q2)*(q3-q2) + (q4-q1)*(q4-q1));
+	G(1,0) = 2 * q4 / sqrtf(1 - 4 * (q2*q3 + q1*q4)*(q2*q3 + q1*q4));
+	G(1,1) = 2 * q3 / sqrtf(1 - 4 * (q2*q3 + q1*q4)*(q2*q3 + q1*q4));
+	G(1,2) = 2 * q2 / sqrtf(1 - 4 * (q2*q3 + q1*q4)*(q2*q3 + q1*q4));
+	G(1,3) = 2 * q1 / sqrtf(1 - 4 * (q2*q3 + q1*q4)*(q2*q3 + q1*q4));
+	G(2,0) = G(0,3);
+	G(2,1) = G(0,2);
+	G(2,2) = G(0,1);
+	G(2,3) = G(0,0);
+
+	return G * quaternion_covariances() * G.transpose();
+}
+
+// get the full covariance matrix
+const matrix::SquareMatrix<float, 24> Ekf::covariances() const
```

Comment:
Did you try using `_k_num_states` instead of 24 here?

Reply:
Yes and got a compile error stating it is undefined.

Reply:
A 321 Tait-Bryan rotation sequence will be badly conditioned around +-90 degrees pitch. The EKF is designed to work for arbitrary rotations, so where we use Euler angles inside the EKF, we also provide an alternative 312 sequence to use when it is better conditioned.

Also I haven't gone through the maths but there is no explicit protection against negative sqrt and /0 floating point exceptions. When I tried using these equations in a Matlab script (attached), the G matrix blew up when RPY was set to zero (quaternion 1,0,0,0)

[test of tait bryan jacobian .m.txt](https://github.com/PX4/ecl/files/2732257/test.of.tait.bryan.jacobian.m.txt)

Reply:
> Also I haven't gone through the maths but there is no explicit protection against negative sqrt and /0 floating point exceptions.

@priseborough how do you suggest to add this protection?

> When I tried using these equations in a Matlab script (attached), the G matrix blew up when RPY was set to zero (quaternion 1,0,0,0)

I will go through the math again to check all's good and there's no typo. If it matches the source, then something else is blowing up the G matrix.

Reply:
@priseborough I have put the source on the comment. Maybe you can take a look at it and assert yourself about the source. I cannot tell myself that the source is correct, it was just the only source I have found that gave me a solution for this aspect and which seems trustful.

Reply:
Also, architecturaly speeking, we do need to propagate the covariances for euler angles on the odometry pipeline, since the message definitions and conventions we currently use on PX4 (and ROS) for odometry oblige to so. We may question about the source but we do need a solution for this in the end.

Reply:
So you don't have to look for it: www.ucalgary.ca/engo_webdocs/GL/96.20096.JSchleppe.pdf